docs/howto/write-and-structure-charm-code.md

```diff
@@ -0,0 +1,307 @@
+structure for building a charm.
+
+```{note}
+In most cases, you'll want to use `--profile=machine` or `profile=kubernetes`.
+If you are charming an application built with a popular framework, check if
+charmcraft has a {external+charmcraft:ref}`specific profile <tutorial>` for it.
+
+Avoid the default (`--profile=simple`), which provides a demo charm, rather than
+a base for building a charm of your own.
+```
+
+If your repository will hold multiple charms, or a charm and source for other
+artifacts, like a Rock, create a `charms` folder at the top level, then a folder
+for each charm inside of that one, and run `charmcraft --init` in each charm
+folder. You'll end up with a structure similar to:
+
+```
+my-charm-set-operators/
+├── charms
+│   ├── my-charm-core
+│   │   ├── charmcraft.yaml
+│   │   ├── pyproject.toml
+│   │   ├── README.md
+│   │   ├── requirements.txt
+│   │   ├── src
+│   │   │   └── charm.py
+│   │   ├── tests
+|   |   |   └── ...
+│   │   └── tox.ini
+│   ├── my-charm-dashboard
+|   |   └── ...
+│   └── my-charm-helper
+|   |   └── ...
+├── CONTRIBUTING.md
+├── LICENSE
+├── README.md
+└── rock
+    └── ...
+```
+
+(define-the-required-dependencies)=
+## Define the required dependencies
+
+### Use the Python provided by the base
+
+Charms run using the Python version provided by the base Ubuntu version. Write
+charm code that will run with the Python version of the oldest base you support.
+
+> See also: {external+juju:ref}`Juju | Roadmap and releases <juju-roadmap-and-releases>`
+
+```{admonition} Best practice
+:class: hint
+
+Set the [`requires-python`](https://packaging.python.org/en/latest/specifications/pyproject-toml/#requires-python)
+version in your `pyproject.toml` so that tooling will detect any use of Python
+features not available in the versions you support.
+```
+
+### Add Python dependencies to pyproject.toml, and generate a lock file
+
+Specify all the direct dependencies of your charm in your `pyproject.toml`
+file in the top-level charm folder. For example:
+
+```toml
+# Required group: these are all dependencies required to run the charm.
+dependencies = [
+    "ops~=2.19",
+]
+
+# Required group: these are all dependencies required to run all the charm's tests.
+[dependency-groups]
+test = [
+    "ops[testing]",
+    "pytest",
+    "coverage[toml]",
+    "jubilant",
+]
+# Optional additional groups:
+docs = [
+    "canonical-sphinx-extensions",
+    "furo",
+    "sphinx ~= 8.0.0",
+    "sphinxext-opengraph",
+]
+```
+
+```{tip}
+Including an external dependency is a significant choice. It can help with
+reducing the complexity and development cost. However, it also increases the
+complexity of understanding the entire system, and adds a maintenance burden of
+keeping track of upstream versions, particularly around security issues.
+
+> See more: [Our Software Dependency Problem](https://research.swtch.com/deps)
+```
+
+Use the `pyproject.toml` dependencies to specify *all* dependencies (including
```

Comment:
This section will get significantly smaller and simpler once we move the charmcraft profiles over to use the `uv` plug-in (we have this as an item for 25.10). At the moment there's a lot of admonition blocks in a row, but I suggest we leave it this way until we can clean it up by just saying more-or-less "use the `uv` plug-in".